### PR TITLE
Statistics [REMOVE PLS A NEW IS OPEN]

### PR DESCRIPTION
**Pull request by ToufouMaster**
Basic Statistics System

### CODE REVIEW

**Review by DeltaEvo — CHANGES_REQUESTED** [review]
You branch contains also changes from your dialog pull request
We are only interested in the last commit
To fix that you can run
```sh
# We add a remote to fetch the data from the upstream repository
git remote add up https://github.com/aresrpg/aresrpg.git  
# Download data from upstream
git fetch up
# Reset the branch at the version from the repository
# This makes you "looses" all your changes, so be careful
git reset --hard up/master
# Reapply your Statistics commit
git cherry-pick 9debb450858d6ccfcf6618afb60d2db1c7bbdb0e
```

**Comment by ToufouMaster** [comment]
Can you explain me how to resolve that pls Thx

**Comment by DeltaEvo** [comment]
I already wrote instructions on how to resolve that

**Comment by ToufouMaster** [comment]
> 
> 
> I already wrote instructions on how to resolve that

i did this but it does not work

**Comment by DeltaEvo** [comment]
Does not work doesn't give me much information, can you be more precise ?

If I take my repo and execute each command I sent, I end with the expected result

**Comment by ToufouMaster** [comment]
ok i retry this later

**Comment by DeltaEvo** [comment]
There is still an issue with this pull request, a lot of files are modified

File: src/commands/commands.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        message: JSON.stringify({
+          translate: 'commands.help.failed',
+          color: 'red',
+        }),
+        position: Position.CHAT,
+        sender: sender.uuid,
+      })
+      break
+  }
+}
```
Comment (Sceat) [inline src/commands/commands.js:73]: This part seems verbose and ambiguous to me

### Sending
It should be consistent, if you send packets through `write_chat_msg` from external imports it should stay that way for the `default` too

### Matching
The `/` part is redundant, you can trim it and then directly apply the command option to the function name
ex:
```js
import command_error from '..'

const available_commands = ['like', 'dislike']
const to_entry = async cmd => [cmd, (await import(`./${cmd}.js`)).default]
const commands_entries = await Promise.all(available_commands.map(to_entry))
const commands = new Map(commands_entries) // ['cmd', function]

commands.set('gm', commands.get('gamemode')) // handling aliases

export default function execute_command({ server, message, sender }) {
  const [name, ...args] = message.trimStart().split(/\s+/)
  const options = {
    name,
    args,
    sender,
    server,
  }
  
  const command = commands.get(name.substring(1))
  if(!command) command_error()
  else command(options)
}
```

For the sake of readability you can be more verbose and directly use

```js
import like from '..'

const commands = new Map([
	['like', like],
	..
])
```

File: src/commands/gamemode.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  sender.write('game_state_change', {
+    reason: 3, // game_state_change is not only for gamemode but for a lot of things. And 3 is for gamemode.
+    gameMode,
+  })
+
+  sender.write('chat', {
+    message: JSON.stringify(message),
+    position: Position.CHAT,
+    sender: '00000000000000000000000000000000',
```
Comment (Sceat) [inline src/commands/gamemode.js:94]: Could this weird magic be a constant sir ?

File: src/mobs/path.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    const new_path = await Promise.race([
+      setTimeoutPromise(next_time - time, false),
+      next.then(() => true),
+    ])
+
+    if (new_path) break
+  }
+
+  yield* path_to_positions(stream, next)
+}
```
Comment (Sceat) [inline src/mobs/path.js:52]: setTimeout has a promise version in node `15.5`

**Review by Sceat — APPROVED** [review]

**Review by Sceat — CHANGES_REQUESTED** [review]
image

File: src/commands/gamemode.js (hunk 2 above)
Comment (Sceat) [reply src/commands/gamemode.js:94]: is it your bitcoin address ?

File: src/chat.js [hunk 4]
```diff
@@ -17,20 +50,41 @@ export function write_chat_msg(
...
+    const { message } = packet
+    if (is_command_function(message)) {
+      log.debug({ sender: client.uuid, command: message }, 'Command')
+      execute_command({ server, message, sender: client })
+      return
+    } else {
+      log.debug({ sender: client.uuid, message }, 'Message')
+    }
+    const formatted_message = message.split(/(%item\d%)/).map((part) => {
+      if (part.match(/(%item\d%)/)) {
```
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/chat.js:64]: This code is already on master and a regex contain state this could break this code

File: src/commands/commands.js (hunk 1 above)
Comment (DeltaEvo) [reply src/commands/commands.js:73]: Already on master and `write_chat_msg` sends a message to all players, `sender.write` send it only to one player, we should rename `write_chat_msg`

File: src/mobs/path.js (hunk 3 above)
Comment (DeltaEvo) [reply src/mobs/path.js:52]: Yep but we use node 14 for now, I will wait for node 16 to be on a LTS

File: src/commands/commands.js (hunk 1 above)
Comment (DeltaEvo) [reply src/commands/commands.js:73]: cc @Zekirax (author of this code)

Comment (DeltaEvo) [reply src/commands/commands.js:73]: For my part I like this switch, it's "simple"

File: src/commands/gamemode.js (hunk 2 above)
Comment (DeltaEvo) [reply src/commands/gamemode.js:94]: cc @Zekirax

File: src/chat.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/chat.js:64]: cc @Zekirax

[the remark this answers is not in the document]
Comment (Vahelnir) [reply src/chat.js:64]: aren't regexes without the g or y flag stateless ?

File: src/commands/gamemode.js (hunk 2 above)
Comment (Zekirax) [reply src/commands/gamemode.js:94]: It's my bank account password.

File: src/mobs/path.js (hunk 3 above)
Comment (Sceat) [reply src/mobs/path.js:52]: die potato

File: src/chat.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (DeltaEvo) [reply src/chat.js:64]: Mmh maybe

**Pull request closed without merging**